### PR TITLE
Add a `doc_paragraphs_missing_punctuation` lint

### PR DESCRIPTION
**Pull request by AudaciousAxiom**
This introduces a new lint that aims to check for missing terminal punctuation in doc comments.

## Lint scope and motivation

It partially addresses https://github.com/rust-lang/rust-clippy/issues/8371 by currently focusing on the case that is both the easiest to look for without advanced NLP capacities and still very useful: missing punctuation at the end of the last sentence of doc comments. This is particularly useful when working in a team to enforce a style guide and make sure all doc comments end with some kind of punctuation, which is often forgotten in the case of short, single-sentence doc comments. The lint is biased towards avoiding false positives so it can be easily adopted without requiring an excessive number of `#[expect]` attributes.

It is currently only concerned with Latin languages, but adding support for `。` for instance should be very easy.

Even if consistently ending doc comments (even very short ones) with punctuation is somewhat of an opinion, it seems sufficiently well-established, [at least in the `std`](https://github.com/rust-lang/rust/pull/91886), to warrant its own lint, even if of course it would be allow-by-default.

## Lint category and possible evolution

Because it is unclear how useful and more complex it would be to also look for missing punctuation at the end of *each* Markdown paragraphs, I opted for the `nursery` category for now. My understanding of [the stability guarantees](https://github.com/rust-lang/rfcs/blob/master/text/2476-clippy-uno.md#stability-guarantees) is that would it not be acceptable to add such capability later if it was part of `pedantic` or `restriction` categories right away. I tried to make sure the lint name allows this kind of iterative evolution.

## Testing

I ran the lint against the `core` library which led me to introduce a few special cases (which seem common enough) to avoid false positives: e.g., the last sentence being in parentheses. After excluding specific modules which should likely be excluded (e.g., `core_arch`, `intrinsincs`), it currently finds a bit more than 200 errors in the `core` library.

This lint also helped find issues with a few doc comments, which [are now fixed](https://github.com/rust-lang/rust/pull/146136).

## How to review this PR

I suppose it is easier to have a look at the UI tests first before checking the implementation.

---

*Please write a short comment explaining your change (or "none" for internal only changes)*

changelog: add new lint [`doc_paragraphs_missing_punctuation`]

### CODE REVIEW

**Comment by AudaciousAxiom** [comment]
Are the lint warnings found by the lintcheck check something that should be addressed in any way or is it just to get an idea of what it looks like to run the new lint on popular crates?

**Comment by notriddle** [comment]
This seems hard to maintain. I would prefer if it was based on the actual markdown parser instead of doing it like this.

I've rewritten a version, that you can see here: https://github.com/AudaciousAxiom/rust-clippy/compare/feat/doc_comments_missing_terminal_punctuation...notriddle:rust-clippy:feat/doc_comments_missing_terminal_punctuation

If it looks better to everyone else, we can push it to your branch?

File: clippy_lints/src/doc/doc_comments_missing_terminal_punctuation.rs [hunk 1]
```diff
@@ -0,0 +1,85 @@
...
+                text_offset = None;
+            },
+            Event::InlineHtml(_) | Event::Start(Tag::Image { .. }) | Event::End(TagEnd::Image) => {
+                text_offset = None;
+            },
+            Event::Start(Tag::Link { .. }) | Event::End(TagEnd::Link) if no_report_depth == 0 && !offset.is_empty() => {
+                text_offset = Some(offset.end);
+            },
+            Event::Text(..) if no_report_depth == 0 && !offset.is_empty() => {
+                if doc_string[..offset.end].trim_end().ends_with(')') {
```
Comment (AudaciousAxiom) [inline clippy_lints/src/doc/doc_comments_missing_terminal_punctuation.rs]: @notriddle This now only checks for parentheses, I'm not sure why you were trimming brackets and curly braces as well?

File: clippy_lints/src/doc/doc_comments_missing_terminal_punctuation.rs [hunk 2]
```diff
@@ -0,0 +1,85 @@
...
+                no_report_depth += 1;
+            },
+            Event::End(TagEnd::FootnoteDefinition) => {
+                no_report_depth -= 1;
+            },
+            Event::End(
+                TagEnd::CodeBlock | TagEnd::Heading(_) | TagEnd::HtmlBlock | TagEnd::List(_) | TagEnd::Table,
+            ) => {
+                no_report_depth -= 1;
+                text_offset = None;
```
Comment (AudaciousAxiom) [inline clippy_lints/src/doc/doc_comments_missing_terminal_punctuation.rs]: I've split this arm so `text_offset` gets overwritten if these constructs are at the end of the doc comments (and added some tests for these cases).

File: tests/ui/doc/doc_paragraphs_missing_punctuation.rs [hunk 3]
```diff
@@ -0,0 +1,159 @@
...
+ *
+ */
+//~^^^ doc_comments_missing_terminal_punctuation
+struct BlockDocComment;
+
+/// Sometimes a doc attribute is used for concatenation
+/// ```
+#[doc = ""]
+/// ```
+struct DocAttribute;
```
Comment (AudaciousAxiom) [inline tests/ui/doc/doc_paragraphs_missing_punctuation.rs:172]: I've opted for simply ignoring interleaved `#[doc]` attributes to reduce false positives, and removed the `_unfixable` test file.

File: tests/ui/doc/doc_comments_missing_terminal_punctuation.fixed [hunk 4]
```diff
@@ -0,0 +1,159 @@
...
+mod module {
+    //! Works on
+    //! inner attributes too.
+    //~^ doc_comments_missing_terminal_punctuation
+}
+
+enum Trailers {
+    /// (Sometimes the last sentence is in parentheses, and that's ok.)
+    ParensPassing,
+    /// (But sometimes it is missing a period.)
```
Comment (AudaciousAxiom) [inline tests/ui/doc/doc_comments_missing_terminal_punctuation.fixed]: This case is now properly handled: the period is added inside the parentheses instead of outside.

File: clippy_lints/src/doc/doc_comments_missing_terminal_punctuation.rs [hunk 5]
```diff
@@ -0,0 +1,85 @@
...
+            },
+            Event::InlineHtml(_) | Event::Start(Tag::Image { .. }) | Event::End(TagEnd::Image) => {
+                text_offset = None;
+            },
+            Event::Start(Tag::Link { .. }) | Event::End(TagEnd::Link) if no_report_depth == 0 && !offset.is_empty() => {
+                text_offset = Some(offset.end);
+            },
+            Event::Code(..) | Event::Text(..) if no_report_depth == 0 && !offset.is_empty() => {
```
Comment (AudaciousAxiom) [inline clippy_lints/src/doc/doc_comments_missing_terminal_punctuation.rs]: Looking at the output of the lintcheck check, I realized that trailing code spans were not properly treated (the lint was suggesting to add a period just before the code span, instead of after it). Adding `Event::Code` here fixes that.

File: clippy_lints/src/doc/doc_comments_missing_terminal_punctuation.rs (hunk 1 above)
Comment (notriddle) [reply clippy_lints/src/doc/doc_comments_missing_terminal_punctuation.rs]: I don't actually remember what my reasoning was, either...

File: tests/ui/doc/doc_comments_missing_terminal_punctuation.fixed (hunk 4 above)
Comment (notriddle) [reply tests/ui/doc/doc_comments_missing_terminal_punctuation.fixed]: Consider this example:

```rust
/// Perform a binary search, returning the first element of the list
/// that is equal (Ok) or the position where the element should be
/// inserted (Err)
fn binary_search<T: Eq>(slice: &[T], elem: T) -> Result<usize, usize> {
    unimplemented!();
}
```

"(Err)" is not a complete sentence, so it should be "(Err).", not "(Err.)".

The more complex solution should be used if it's more common, but is it? Are parenthesis more commonly used for sentence fragments, or for whole sentences?

Comment (AudaciousAxiom) [reply tests/ui/doc/doc_comments_missing_terminal_punctuation.fixed]: > The more complex solution should be used if it's more common, but is it?

Clearly it isn't. I've considered *also* accepting as valid comments that end with `.)`, but realized that would result in false positives for comments that end with `etc.)` for instance (and just in the std there are three of them).

So after all I've completely removed any special treatment of parentheses (and updated the tests to reflect this). If you think it would be better to also accept `.)` to avoid some false positives, I can re-add that easily.

(I've dropped the commit that was adding the special treatment of parens, I'm not sure what the stance is on rewriting history in PRs, but in any case I can easily bring it back later if needed.)

Comment (notriddle) [reply tests/ui/doc/doc_comments_missing_terminal_punctuation.fixed]: I just thought of another corner case...

```rust
/// Says "yes."
pub fn affirm()
```

Putting the period inside the quotes doesn't make sense, but it's [standard US English orthography](https://en.wikipedia.org/wiki/Quotation_marks_in_English#Order_of_punctuation), so we don't want to warn on it. British orthography puts the full stop outside the quotes, the same way brackets work, so we need to support both.

Whatever solution you use for quotes-with-`.`-inside can probably also be used for parens.

Comment (AudaciousAxiom) [reply tests/ui/doc/doc_comments_missing_terminal_punctuation.fixed]: I think I actually thought of that one before but decided against as I thought it was just too rare to actually happen in doc comments and especially at their very end. Out of curiosity I just grepped the std code and only found [this one occurrence](https://github.com/rust-lang/rust/blob/a2db9280539229a3b8a084a09886670a57bc7e9c/library/core/src/slice/memchr.rs#L16). Do you think this is worth the additional complexity?

Comment (notriddle) [reply tests/ui/doc/doc_comments_missing_terminal_punctuation.fixed]: I'm not sure how common it is. Maybe [lintcheck](https://doc.rust-lang.org/clippy/development/basics.html?highlight=lintcheck#lintcheck) can be used to try to find out?

Comment (AudaciousAxiom) [reply tests/ui/doc/doc_comments_missing_terminal_punctuation.fixed]: Accepting doc comments that end with double quotation marks (and with proper punctuation inside them) brought the linkcheck count from 6669 down to 6614. I don't think the *slight* additional complexity is an issue actually.

File: clippy_lints/src/doc/doc_comments_missing_terminal_punctuation.rs [hunk 6]
```diff
@@ -0,0 +1,88 @@
...
+                text_offset = None;
+            },
+            Event::Code(..) | Event::Start(Tag::Link { .. }) | Event::End(TagEnd::Link)
+                if no_report_depth == 0 && !offset.is_empty() =>
+            {
+                text_offset = Some(offset.end);
+            },
+            Event::Text(..) if no_report_depth == 0 && !offset.is_empty() => {
+                // American-style quotes require punctuation to be placed inside closing quotation marks.
+                if doc_string[..offset.end].trim_end().ends_with('"') {
```
Comment (notriddle) [inline clippy_lints/src/doc/doc_comments_missing_terminal_punctuation.rs]: Okay, so this code can be extended to handle parens with basically-trivial additional complexity, right?


```suggestion
                if doc_string[..offset.end].trim_end().ends_with(|c| c == '"' || c == ')') {
```

Comment (AudaciousAxiom) [reply clippy_lints/src/doc/doc_comments_missing_terminal_punctuation.rs]: Sure, but that means that the fix suggestion would always add a period inside the parentheses, while it is more likely that adding it outside would be the correct choice. That makes me realize that the suggestion for quotes would also arbitrarily favor the American style over the British one.

Maybe we should treat these cases (quotes and parentheses) as unfixable: detect when terminal punctuation is missing but give no suggestions?

Comment (AudaciousAxiom) [reply clippy_lints/src/doc/doc_comments_missing_terminal_punctuation.rs]: To be honest I think we are overfocusing on edge cases right now: this lint (even without handling quotes and parentheses) would already be really useful for rolling out on new projects to ensure a consistent style, and we know that we will not be able to handle every little edge case without full-blown NLP anyway. These edge cases can easily be circumvented by simply slightly rephrasing (and I understand that some people do not like tools dictating how they write, but in this case this seems pretty minor).

Comment (notriddle) [reply clippy_lints/src/doc/doc_comments_missing_terminal_punctuation.rs]: > Maybe we should treat these cases (quotes and parentheses) as unfixable: detect when terminal punctuation is missing but give no suggestions?

Yes, I agree. That would be great!

Comment (notriddle) [reply clippy_lints/src/doc/doc_comments_missing_terminal_punctuation.rs]: > To be honest I think we are overfocusing on edge cases right now: this lint (even without handling quotes and parentheses) would already be really useful for rolling out on new projects to ensure a consistent style, and we know that we will not be able to handle every little edge case without full-blown NLP anyway.

It's certainly true that we can't reach perfection without full AI. When we inevitably err, though, I'd prefer to err on the side of false negatives instead of false positives. We could even solve the parens and quotes problem by adding them as accepted terminating punctuation:

```rust
const TERMINAL_PUNCTUATION_MARKS: &[char] = &['.', '?', '!', '…', ')', '"', '\''];
```

The point is that we don't want to make people `#[allow]` the lint.

Comment (AudaciousAxiom) [reply clippy_lints/src/doc/doc_comments_missing_terminal_punctuation.rs]: I've pushed an update that does the following:

- Consider both `.)` and `).` to be correct (thus there could be false negatives in
  the former case, but nothing we can do about it).
- Consider both `."` and `".` to be correct (same).
- Otherwise the lint triggers and offers help, but does not suggest a fix.

The implementation could likely use some refactoring, but we should agree on the
behavior before that.

I did not modify `TERMINAL_PUNCTUATION_MARKS` because that was causing issues with Markdown links, which also end with a parenthesis. I also did not do anything about single quotation marks, because my understanding is that those are used in British English and I would expect American style not be used in that case (and therefore no special treatment is needed).

File: clippy_lints/src/doc/doc_paragraphs_missing_punctuation.rs [hunk 7]
```diff
@@ -0,0 +1,117 @@
...
+        IsMissingPunctuation::No => {},
+    }
+}
+
+#[must_use]
+/// If punctuation is missing, returns the offset where new punctuation should be inserted.
+fn is_missing_punctuation(doc_string: &str) -> IsMissingPunctuation {
+    const TERMINAL_PUNCTUATION_MARKS: &[char] = &['.', '?', '!', '…'];
+
+    let mut no_report_depth = 0;
```
Comment (Alexendoo) [inline clippy_lints/src/doc/doc_paragraphs_missing_punctuation.rs:53]: Regarding catching individual paragraphs, would it suffice to track the unconditional depth of start/end tags and check when it goes from 1 -> 0 due to a `TagEnd::Paragraph`?

File: clippy_lints/src/doc/doc_paragraphs_missing_punctuation.rs [hunk 8]
```diff
@@ -0,0 +1,117 @@
...
+    let mut missing_punctuation = IsMissingPunctuation::No;
+    for (event, offset) in
+        Parser::new_ext(doc_string, main_body_opts() - Options::ENABLE_SMART_PUNCTUATION).into_offset_iter()
```
Comment (Alexendoo) [inline clippy_lints/src/doc/doc_paragraphs_missing_punctuation.rs:58]: Double parsing every doc comment is not ideal, could this be tweaked into a state machine struct that `check_doc` passes the individual events into?

File: clippy_lints/src/doc/doc_paragraphs_missing_punctuation.rs (hunk 7 above)
Comment (notriddle) [reply clippy_lints/src/doc/doc_paragraphs_missing_punctuation.rs:53]: Unfortunately, the "rules" for what makes sense in the middle of a comment are different from the rules that make sense at the end. For example, here's a few crates where a paragraph ends in a colon:

- <https://docs.rs/regex/latest/regex/struct.RegexBuilder.html#method.unicode> "Enabling Unicode does a number of things:"
- <https://doc.rust-lang.org/nightly/std/collections/index.html> "Rust’s collections can be grouped into four major categories:"
- <https://docs.rs/arti/1.5.0/arti/index.html#using-arti-with-tor-browser> "After that launch it with some basic configuration parameters:"

Comment (Alexendoo) [reply clippy_lints/src/doc/doc_paragraphs_missing_punctuation.rs:53]: I don't see that being a problem, it would no longer lint on a final paragraph ending in `:` but that doesn't seem like it would be common enough to need a lint

Comment (AudaciousAxiom) [reply clippy_lints/src/doc/doc_paragraphs_missing_punctuation.rs:53]: I've finally gotten around to implement this: as you suggested, each paragraph end is now checked, and colons are allowed in that position. I've of course updated the tests to reflect that. I also had to refactor a bit to be able to return multiple lint error for each doc comment.

Do you think the lint should be renamed, maybe by removing the word "terminal", because colons are not technically terminal punctuation (and that would make the name shorter)?

(As rustbot shows, I had to rebase to take into account the new way of importing `pulldown_cmark`.)

File: clippy_lints/src/doc/doc_paragraphs_missing_punctuation.rs (hunk 8 above)
Comment (AudaciousAxiom) [reply clippy_lints/src/doc/doc_paragraphs_missing_punctuation.rs:58]: Before adding support for checking each paragraph, I had added a short-circuit in cases where punctuation was obviously correct. With this new check this is not possible anymore however.

What are the performance implications when the lint is *not* enabled by users? Does it still run? It is my impression that this used to be the case but is not anymore. If not, do you really want this in this PR, or can this be left for a separate refactor, to avoid modifying multiple lints at once?

File: clippy_lints/src/doc/doc_paragraphs_missing_punctuation.rs (hunk 7 above)
Comment (Alexendoo) [reply clippy_lints/src/doc/doc_paragraphs_missing_punctuation.rs:53]: Yeah that sounds good, maybe also specify that it's about paragraphs now with something like `doc_paragraphs_missing_punctuation`

File: clippy_lints/src/doc/doc_paragraphs_missing_punctuation.rs (hunk 8 above)
Comment (Alexendoo) [reply clippy_lints/src/doc/doc_paragraphs_missing_punctuation.rs:58]: It runs as long as any of the lints in the `Documentation` pass are enabled (warn or above) at any point in the current crate, since many are on by default this means pretty much always

Yeah sure it can be deferred, I just spotted that there's another lint that also separately parses the comments in this module so we could do them both at once

**Comment by Alexendoo** [comment]
FCP opened - https://rust-lang.zulipchat.com/#narrow/channel/257328-clippy/topic/FCP.20doc_comments_missing_terminal_punctuation/with/545833836

File: clippy_lints/src/doc/doc_paragraphs_missing_punctuation.rs (hunk 7 above)
Comment (AudaciousAxiom) [reply clippy_lints/src/doc/doc_paragraphs_missing_punctuation.rs:53]: Agreed, I've renamed the lint: every occurrence in source code, the changelog entry in the PR description, and the PR title. We unfortunately can't rename the branch.

I think this is now ready for review; tell me if you want me to squash the commits (because their titles barely make sense now) while of course properly preserving authorship.

**Comment by AudaciousAxiom** [comment]
As the FCP has been open for more than two weeks, I've updated the lint metadata: I've switched `clippy::version` to 1.93.0 and set the category to `restriction`.

**Comment by AudaciousAxiom** [comment]
However when setting the category to `restriction` we get an unrelated lint error in `tests/ui/blanket_clippy_restriction_lints.rs`: would it be ok to simply add a period at the end of the doc comment in that file?

File: clippy_lints/src/doc/doc_paragraphs_missing_punctuation.rs [hunk 9]
```diff
@@ -0,0 +1,126 @@
...
+        match missing_punctuation {
+            MissingPunctuation::Fixable(offset) => {
+                // This ignores `#[doc]` attributes, which we do not handle.
+                if let Some(span) = fragments.span(cx, offset..offset) {
+                    clippy_utils::diagnostics::span_lint_and_sugg(
+                        cx,
+                        DOC_PARAGRAPHS_MISSING_PUNCTUATION,
+                        span,
+                        MSG,
+                        "end the doc comment with some punctuation",
```
Comment (Alexendoo) [inline clippy_lints/src/doc/doc_paragraphs_missing_punctuation.rs]: This should also say paragraph

**Review by Alexendoo — COMMENTED** [review]
Thanks, yeah just adding the `.` there is fine

Some cases that may be common from looking through the lintcheck - 

A final parenthetical https://docs.rs/ahash/0.8.11/src/ahash/lib.rs.html#251

A link after a colon: https://docs.rs/aho-corasick/1.1.3/src/aho_corasick/packed/rabinkarp.rs.html#30-34

Without being too clever we could consider `)` trailing punctuation, could miss some cases where it is desired to add a `.`, but also sidesteps the issue of where exactly it should go

For URLs I'm not sure, adding `.` doesn't break the URL but does look a little strange to me

*[View changes since this review](https://triagebot.infra.rust-lang.org/gh-changes-since/rust-lang/rust-clippy/15758/1ecb18a18211c86a2a2f00ef3b0ef4c54f54d0d2..d6936cbd8cf17088858d8e8390f6250492f528cb)*

File: clippy_lints/src/doc/doc_paragraphs_missing_punctuation.rs (hunk 9 above)
Comment (AudaciousAxiom) [reply clippy_lints/src/doc/doc_paragraphs_missing_punctuation.rs]: Indeed, I've changed "doc comment" to "paragraph" (to keep the message short), and updated the tests.

**Comment by AudaciousAxiom** [comment]
I agree that we should strive for a low rate of false positives, but regarding these two specific cases (`ahash/0.8.11/src/ahash/lib.rs.html#251` and `1.1.3/src/aho_corasick/packed/rabinkarp.rs.html#30-34`) I do think these should be errors:

- Parentheses already have special handling in this lint, and I don't think `(Sealed)` should be considered correct (mostly because that's not a sentence). There are multiple ways of making this pass the lint, either by simply adding a period in the parentheses (`(Sealed.)`, so that it's a complete sentence), writing more words (e.g., `(This trait is sealed.)`), or simply rephrasing the comment. In addition, I really don't think we should consider a closing parenthesis terminal punctuation, because spotting missing periods immediately following a parenthesis at the end of a sentence is definitely one of the goal of this lint (see the `UnfixableTrailers::EndsWithParens` test case).
  Concerning the suggestion in the case of parenthesis, we've realized above that we wouldn't be able to provide correct-enough fixes, and decided to *detect* these cases of missing punctuation, but not provide a suggestion (i.e., treat these as unfixable), so the author can decide where the punctuation should go (these should be rare enough).
- Regarding links at the end of a sentence: to me that's really the kind of thing that *should* be spotted by this lint. That occurence should be rewritten by enclosing the link in angled brackets, and adding a period to end the sentence (`{first clause}: <{autolink}>.`).

**Comment by AudaciousAxiom** [comment]
@notriddle @Alexendoo Thanks a lot for your help and reviews!
